### PR TITLE
Refactor RL model wrapper into a `trainer` module

### PR DESCRIPTION
**Pull request by jon-tow**
This PR refactors the RL model wrappers into "trainer" wrappers. The term "model" has semantic overloading throughout the codebase. A specific point of confusion is the `{Type}RLModel`s, which do not only contain models but also wrap around optimizers, schedulers, and other auxiliary data structures required for RL training.

This refactor was briefly discussed in the CarperAI Discord with @cat-state. I am leaving it here as a reminder and for others to chime in.

### CODE REVIEW

**Comment by LouisCastricato** [comment]
I am strongly in favor of this refactor

File: tests/test_configs.py [hunk 1]
```diff
@@ -31,6 +31,8 @@ def test_repo_trl_configs():
         assert os.path.isfile(file), f"Config file {file} does not exist."
         assert file.endswith(".yml"), f"Config file {file} is not a yaml file."
         try:
-            TRLConfig.load_yaml(file)
+            config = TRLConfig.load_yaml(file)
+            assert config.train.entity_name is None, \
```
Comment (LouisCastricato) [inline tests/test_configs.py:35]: Thank god lmao

File: trlx/trainer/__init__.py [hunk 2]
```diff
@@ -11,17 +11,17 @@
...
 
 # specifies a dictionary of architectures
-_MODELS: Dict[str, Any] = {}  # registry
+_TRAINERS: Dict[str, Any] = {}  # registry
 
 
-def register_model(name):
-    """Decorator used register an architecture
+def register_trainer(name):
+    """Decorator used to register a trainer
```
Comment (LouisCastricato) [inline trlx/trainer/__init__.py:18]: unrelated but a lot of these comment strings havent been changed since they were copied over from magiCARP ;)

**Review by LouisCastricato — APPROVED** [review]

File: examples/architext.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 # Toy example of optimizing textual interior designs to output the least number of rooms
 # Also see https://architext.design/
-
+import yaml
 import trlx
+from trlx.data.configs import TRLConfig
```
[the remark this answers is not in the document]
Comment (jon-tow) [reply examples/architext.py:5]: `TRLConfig` and `yaml` were never defined.

File: trlx/trainer/__init__.py (hunk 2 above)
Comment (jon-tow) [reply trlx/trainer/__init__.py:18]: 🎣

File: examples/experiments/grounded_program_synthesis/train_trlx.py [hunk 4]
```diff
@@ -59,12 +59,12 @@ def main(hparams={}):
...
     train_prompts = list(dataset.load_datapoints(split="train"))[:1000]
 
-    model = trlx.train(
+    trainer = trlx.train(
         reward_fn=reward_fn,
         prompts=train_prompts,
         config=config,
     )
-    model.save_pretrained("dataset/trained_model")
+    trainer.save_pretrained("dataset/trained_model")
```
Comment (Dahoas) [inline examples/experiments/grounded_program_synthesis/train_trlx.py:67]: Do the model types we use support save_pretrained?

**Review by Dahoas — APPROVED** [review]
Looks good! Will merge if there are no further changes

File: examples/experiments/grounded_program_synthesis/train_trlx.py (hunk 4 above)
Comment (LouisCastricato) [reply examples/experiments/grounded_program_synthesis/train_trlx.py:67]: Yes

Comment (Dahoas) [reply examples/experiments/grounded_program_synthesis/train_trlx.py:67]: Wait I don't think they do, or at least ppo doesn't.

Comment (LouisCastricato) [reply examples/experiments/grounded_program_synthesis/train_trlx.py:67]: Ohhhh ok wait that's weird then can we just add a save pretrained function to PPO haha

Comment (Dahoas) [reply examples/experiments/grounded_program_synthesis/train_trlx.py:67]: But anyway saving doesn't have anything to do with this pr so I think it's fine for now.

**Pull request merged**